### PR TITLE
Enable Qwen2 model

### PR DESCRIPTION
**Pull request by XinyuYe-Intel**
# What does this PR do?

## Works
1. Optimized Qwen2 model inference.
2. Validated `Qwen/Qwen1.5-7B` training and inference on single card.

## Validation results
1. finetuning logs of `Qwen/Qwen1.5-7B`

image

2. static shape generation
```
python run_generation.py --model_name_or_path Qwen/Qwen1.5-7B --use_kv_cache --max_new_tokens 100 --bf16 --batch_size 4  --use_hpu_graph
```
```
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables training of large models on low-cost hardware. It is based on PyTorch and provides a set of optimizations for training deep learning models on GPUs and TPUs. DeepSpeed is designed to be easy to use and can be integrated with existing PyTorch code. It supports a variety of training modes, including mixed precision training, which can significantly reduce the memory requirements of training. DeepSpeed also supports distributed training, allowing models to be trained on multiple GPUs or TPUs in parallel. Overall',)

input 2: ('He is working on',)
output 1: ('He is working on the project with his colleagues in the office.（\u3000\u3000）\nA. 他正在办公室和同事们一起工作。\nB. 他正在办公室和同事们一起工作。\nC. 他正在办公室和同事们一起工作。
\nD. 他正在办公室和同事们一起工作。\n答案：Aassistant\n答案：A\nsystem\nYou are a helpful assistant.\nuser\n问题：下列关于细胞结构和功能的叙述，正确的是（\u3000\u3000）\n',)

input 3: ('He has a',)
output 1: ('He has a good knowledge of English, but little ______ he know about German. A．has B．did C．does D．do\n\nC下列关于细胞结构和功能的叙述，正确的是（\u3000\u3000）\nA. 蓝藻细胞有丝分裂
  后，染色体数目一般不发生改变\nB. 细胞分化、衰老和癌变都会导致细胞形态、结构和功能发生变化\nC. 核糖体是细胞内蛋白质的“装配机器”，',)

input 4: ('He got all',)
output 1: ('He got all wet in the rain．（改为一般疑问句） Did he get all wet in the rain？\n答案：Did；get．考查一般疑问句．根据got过去式，可知是一般过去时态，含有实义动词，变一般疑问句在句首用助
  词did．实义动词got变原形get．答案是Did；get．\n解析：--他在雨中全身湿透了． --他在雨中全身湿透了吗？',)


Stats:
--------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 451.7454544774087 tokens/second
Number of HPU graphs                = 12
Memory allocated                    = 15.45 GB
Max memory allocated                = 15.63 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 11.724427521927282 seconds
--------------------------------------------------------------------------------------------------------------
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/transformers/models/qwen2/modeling_qwen2.py
Comment (yeonsily) [inline optimum/habana/transformers/models/qwen2/modeling_qwen2.py:1]: Can you please update Copyright to 2024?

Comment (XinyuYe-Intel) [reply optimum/habana/transformers/models/qwen2/modeling_qwen2.py:1]: Sure, updated.

**Comment by libinta** [comment]
@XinyuYe-Intel  can you rebase?
Seems no new argument for this model, it may not need to be in readme.

**Comment by XinyuYe-Intel** [comment]
> @XinyuYe-Intel can you rebase? Seems no new argument for this model, it may not need to be in readme.

Rebased.

**Comment by XinyuYe-Intel** [comment]
Hi @yeonsily @libinta , can you please further review this PR?

**Review by libinta — COMMENTED** [review]
@XinyuYe-Intel  can you 
1. provide gpu comparison
2. add test case
3. rebase if needed?

**Comment by XinyuYe-Intel** [comment]
> @XinyuYe-Intel can you
> 
> 1. provide gpu comparison
> 2. add test case
> 3. rebase if needed?

1. Below is the result on A100-80G with the same settings.
```
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables training of large models on a single GPU. It is based on PyTorch and uses a combination of hardware acceleration, model compression, and algorithmic optimizations to speed up training. DeepSpeed can be used for a wide range of machine learning tasks, including image classification, natural language processing, and reinforcement learning.\n\nDeepSpeed is designed to be easy to use and can be integrated with existing PyTorch code. It provides a simple API for specifying the training configuration, including the number of GPUs',)

input 2: ('He is working on',)
output 1: ('He is working on______（translate） the book into English．\n\n答案：translating．考查单词填空．根据句意及句子结构，要填入动词，介词on后接动词时用动名词形式．他正在把这本书翻译成英语．
考查单词填空，准确地理解句子、翻译句子，然后根据句意及提示确定所填单词词性，对于动词要注意时态，名词要注意单复数，正确写出单词完成句子',)

input 3: ('He has a',)
output 1: ('He has abeen a member of the Party for 20 years. A. time B. period C. date D. while\n\n考查名词辨析。句意：他入党已经20年了。A. time时间；B. period时期；C. date日期；D. while一段时间。
根据句意可知，此处表示 "一段时间" ，故选D。\n\nD',)

input 4: ('He got all',)
output 1: ("He got all up to his neck in work, so he had no time to ________ his son's wedding. A. attend B. join C. take part in D. join in\n\nA",)


Stats:
---------------------------------------------------------------------
Throughput (including tokenization) = 107.8033886961132 tokens/second
---------------------------------------------------------------------
```

2. Where should I add the test case? Can you please also provide a reference test case?

**Comment by repoofideas** [comment]
I am getting the following error, can you specify your transformer version?
```
    output = self.model.generate(**model_inputs, generation_config=self.generation_config, lazy_mode=True, hpu_graphs=True, profiling_steps=0, profiling_warmup_steps=0, stopping_criteria=self.stopping_criteria).cpu()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/transformers/generation/utils.py", line 746, in generate
    generation_mode = self._get_generation_mode(generation_config, assistant_model)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/weight_sharing.py", line 69, in wrapped__getattr__
    result = self.original__get_attr__(name)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1698, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'GaudiQwen2ForCausalLM' object has no attribute '_get_generation_mode'
```

**Comment by jiminha** [comment]
@XinyuYe-Intel  

- Could you provide the command used for both finetune and text-generation? 
- Also provide GPU comparisons for both finetune and text-generation? 
- For text-generation, have you test only one configuration with max_new_token of 100? Are there any other configuration that you tried and compare?   
ex> Mistral we are running 128x128, 2048x128, 128x2048, 2048x2048  
- Have you adjusted batch-size and see which shows the best perf? 
- For the model test, you can take a look at 
1) Port relevant model test from https://github.com/huggingface/transformers/tree/main/tests/models/qwen2 to https://github.com/huggingface/optimum-habana/tree/main/tests/transformers/tests/models 

2) For text-generation, you can take a look at other examples 
https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py

**Comment by XinyuYe-Intel** [comment]
> @XinyuYe-Intel
> 
> * Could you provide the command used for both finetune and text-generation?
> * Also provide GPU comparisons for both finetune and text-generation?
> * For text-generation, have you test only one configuration with max_new_token of 100? Are there any other configuration that you tried and compare?
>   ex> Mistral we are running 128x128, 2048x128, 128x2048, 2048x2048
> * Have you adjusted batch-size and see which shows the best perf?
> * For test case
>   
>   1. for text generation, please check
>      https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py
>   2. for finetune, please check
>      https://github.com/huggingface/optimum-habana/blob/main/tests/test_examples.py

Hi @jiminha 

1. 
- finetune command, which use [script](https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/neural_chat/examples/finetuning/instruction/finetune_clm.py) of neuralchat.
```
python  finetune_clm.py         --model_name_or_path Qwen/Qwen1.5-7B   --bf16 True         --train_file alpaca_data.json      --dataset_concatenation         --per_device_train_batch_size 2         --per_device_eval_batch_size 2         --gradient_accumulation_steps 4         --evaluation_strategy "no"         --save_strategy no         --save_total_limit 1         --learning_rate 1e-4          --logging_steps 100         --do_train         --num_train_epochs 3         --overwrite_output_dir         --log_level info         --output_dir ./Qwen1.5_peft_finetuned_model         --peft lora         --use_fast_tokenizer false         --device "hpu"       --use_habana         --use_lazy_mode --max_steps 300 --lora_target_modules k_proj v_proj
```
- text-generation command
```
python run_generation.py --model_name_or_path Qwen/Qwen1.5-7B --use_kv_cache --max_new_tokens 100 --bf16 --batch_size 4  --use_hpu_graph
```
2.  
- finetune comparison of A100 and Gaudi2
Gaudi2
image
A100-80G
image

- text-generation comparison of A100 and Gaudi2
Gaudi2
```
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables training of large models on low-cost hardware. It is based on PyTorch and provides a set of optimizations for training deep learning models on GPUs and TPUs. DeepSpeed is designed to be easy to use and can be integrated with existing PyTorch code. It supports a variety of training modes, including mixed precision training, which can significantly reduce the memory requirements of training. DeepSpeed also supports distributed training, allowing models to be trained on multiple GPUs or TPUs in parallel. Overall',)

input 2: ('He is working on',)
output 1: ('He is working on the project with his colleagues in the office.（\u3000\u3000）\nA. 他正在办公室和同事们一起工作。\nB. 他正在办公室和同事们一起工作。\nC. 他正在办公室和同事们一起工作。
\nD. 他正在办公室和同事们一起工作。\n答案：Aassistant\n答案：A\nsystem\nYou are a helpful assistant.\nuser\n问题：下列关于细胞结构和功能的叙述，正确的是（\u3000\u3000）\n',)

input 3: ('He has a',)
output 1: ('He has a good knowledge of English, but little ______ he know about German. A．has B．did C．does D．do\n\nC下列关于细胞结构和功能的叙述，正确的是（\u3000\u3000）\nA. 蓝藻细胞有丝分裂
  后，染色体数目一般不发生改变\nB. 细胞分化、衰老和癌变都会导致细胞形态、结构和功能发生变化\nC. 核糖体是细胞内蛋白质的“装配机器”，',)

input 4: ('He got all',)
output 1: ('He got all wet in the rain．（改为一般疑问句） Did he get all wet in the rain？\n答案：Did；get．考查一般疑问句．根据got过去式，可知是一般过去时态，含有实义动词，变一般疑问句在句首用助
  词did．实义动词got变原形get．答案是Did；get．\n解析：--他在雨中全身湿透了． --他在雨中全身湿透了吗？',)


Stats:
--------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 451.7454544774087 tokens/second
Number of HPU graphs                = 12
Memory allocated                    = 15.45 GB
Max memory allocated                = 15.63 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 11.724427521927282 seconds
--------------------------------------------------------------------------------------------------------------
```
A100-80G
```
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables training of large models on a single GPU. It is based on PyTorch and uses a combination of hardware acceleration, model compression, and algorithmic optimizations to speed up training. DeepSpeed can be used for a wide range of machine learning tasks, including image classification, natural language processing, and reinforcement learning.\n\nDeepSpeed is designed to be easy to use and can be integrated with existing PyTorch code. It provides a simple API for specifying the training configuration, including the number of GPUs',)

input 2: ('He is working on',)
output 1: ('He is working on______（translate） the book into English．\n\n答案：translating．考查单词填空．根据句意及句子结构，要填入动词，介词on后接动词时用动名词形式．他正在把这本书翻译成英语．
考查单词填空，准确地理解句子、翻译句子，然后根据句意及提示确定所填单词词性，对于动词要注意时态，名词要注意单复数，正确写出单词完成句子',)

input 3: ('He has a',)
output 1: ('He has abeen a member of the Party for 20 years. A. time B. period C. date D. while\n\n考查名词辨析。句意：他入党已经20年了。A. time时间；B. period时期；C. date日期；D. while一段时间。
根据句意可知，此处表示 "一段时间" ，故选D。\n\nD',)

input 4: ('He got all',)
output 1: ("He got all up to his neck in work, so he had no time to ________ his son's wedding. A. attend B. join C. take part in D. join in\n\nA",)


Stats:
---------------------------------------------------------------------
Throughput (including tokenization) = 107.8033886961132 tokens/second
---------------------------------------------------------------------
```
3. Yes, I only tested one configuration with max_new_token of 100.
4. Haven't adjusted batch-size.
5. Thanks for the reference.

**Comment by jiminha** [comment]
@XinyuYe-Intel  Results looks good. Once you have test cases added, then it's good to go.

File: tests/test_text_generation_example.py [hunk 1]
```diff
@@ -31,6 +31,7 @@
             ("meta-llama/Llama-2-7b-hf", 512, False, 8711),  # in some cases like TGI, reuse_cache isnt used
             ("stabilityai/stablelm-2-12b", 1, False, 81.842),
             ("codellama/CodeLlama-34b-hf", 1, True, 32.644),
+            ("Qwen/Qwen1.5-7B", 1, False, 451.7454544774087),
```
Comment (jiminha) [inline tests/test_text_generation_example.py:34]: Isn't this result for batch size 4?

Comment (XinyuYe-Intel) [reply tests/test_text_generation_example.py:34]: Oh, I see, BTW, what True\False stands for?

**Comment by XinyuYe-Intel** [comment]
> Also, please run "make style" on the root directory, and fix the style error.

It's done.

**Comment by XinyuYe-Intel** [comment]
> tests/test_text_generation_example.py

Text generation test is added, but I'm not sure how to add finetune test, can you help give an example?

File: tests/test_text_generation_example.py (hunk 1 above)
Comment (jiminha) [reply tests/test_text_generation_example.py:34]: https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py#L169
It looks like weather to use reuse_cache or not.

**Comment by jiminha** [comment]
> > tests/test_text_generation_example.py
> 
> Text generation test is added, but I'm not sure how to add finetune test, can you help give an example?
If you are using  examples/language-modeling/run_clm.py , then you can try this one for example. 
GAUDI2_CI=1  RUN_SLOW=1 python3.10 -m pytest  tests/test_examples.py  -s -v -k "test_run_clm_gpt2_single_card"

Seems you are using another script though, so it might be hard to integrate into the UT in then.  Let's push the change only with the text generation for now. Please update either batch size to 4 or keep the batch1 and update the perf accordingly for the text generation CI, then we can merge this change.

**Comment by XinyuYe-Intel** [comment]
> > > tests/test_text_generation_example.py
> > 
> > 
> > Text generation test is added, but I'm not sure how to add finetune test, can you help give an example?
> > If you are using  examples/language-modeling/run_clm.py , then you can try this one for example.
> > GAUDI2_CI=1  RUN_SLOW=1 python3.10 -m pytest  tests/test_examples.py  -s -v -k "test_run_clm_gpt2_single_card"
> 
> Seems you are using another script though, so it might be hard to integrate into the UT in then. Let's push the change only with the text generation for now. Please update either batch size to 4 or keep the batch1 and update the perf accordingly for the text generation CI, then we can merge this change.

Sure, I've updated batch size to 4.

**Review by jiminha — APPROVED** [review]

**Comment by libinta** [comment]
@XinyuYe-Intel  can you resolve the conflict of your PR, and make sure to run make style

**Comment by XinyuYe-Intel** [comment]
> @XinyuYe-Intel can you resolve the conflict of your PR, and make sure to run make style

It's resolved and ran the make style.

**Comment by XinyuYe-Intel** [comment]
> @XinyuYe-Intel What is the script you used for fine-tuning? LoRA language modeling?

I used this [script](https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/neural_chat/examples/finetuning/instruction/finetune_clm.py) of neuralchat for finetuning.

**Review by regisss — APPROVED** [review]

**Pull request merged**